File: src/main/java/com/jozufozu/exnihiloomnia/integration/jei/sieve/SieveRecipeCategory.java
package com.jozufozu.exnihiloomnia.integration.jei.sieve;

import com.jozufozu.exnihiloomnia.ExNihilo;
import com.jozufozu.exnihiloomnia.common.registries.WeightedDrop;
import mezz.jei.api.IGuiHelper;
import mezz.jei.api.gui.IDrawable;
import mezz.jei.api.gui.IGuiItemStackGroup;
import mezz.jei.api.gui.IRecipeLayout;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.recipe.IRecipeCategory;
import net.minecraft.client.Minecraft;
import net.minecraft.client.resources.I18n;
import net.minecraft.util.ResourceLocation;

import javax.annotation.Nonnull;
import java.util.List;

public class SieveRecipeCategory implements IRecipeCategory<SieveRecipeWrapper>
{
    public static final String UID = "exnihiloomnia.sieve";
    
    private static final ResourceLocation texture = new ResourceLocation(ExNihilo.MODID, "textures/gui/jei_sieve.png");
    private static IDrawable background;
    private static IDrawable highlight;
    
    public SieveRecipeCategory(IGuiHelper guiHelper)
    {
        background = guiHelper.createDrawable(texture, 0, 0, 166, 146);
        highlight = guiHelper.createDrawable(texture, 166, 0, 18, 18);
    }
    
    @Nonnull
    @Override
    public String getUid()
    {
        return UID;
    }
    
    @Nonnull
    @Override
    public String getTitle()
    {
        return I18n.format("jei.exnihiloomnia.sieve");
    }
    
    @Nonnull
    @Override
    public String getModName()
    {
        return ExNihilo.NAME;
    }
    
    @Nonnull
    @Override
    public IDrawable getBackground()
    {
        return background;
    }
    
    @Override
    public void drawExtras(Minecraft minecraft)
    {
    
    }
    
    @Override
    public void setRecipe(IRecipeLayout recipeLayout, SieveRecipeWrapper sieveRecipeWrapper, IIngredients ingredients)
    {
        IGuiItemStackGroup itemStacks = recipeLayout.getItemStacks();
        
        itemStacks.init(0, true, 74, 9);
        
        final int outputSlots = 1;
    
        for (int slotNumber = 0; slotNumber < 45; slotNumber++)
        {
            final int slotX = 2 + (slotNumber % 9 * 18);
            final int slotY = 52 + (slotNumber / 9 * 18);
        
            itemStacks.init(outputSlots + slotNumber, false, slotX, slotY);
        }
        
        itemStacks.set(ingredients);
        
        itemStacks.addTooltipCallback((slot, isInput, stack, toolTip) ->
        {
            if (isInput)
                return;
            
            List<WeightedDrop> dropList = sieveRecipeWrapper.rewards.get(stack);
            for (WeightedDrop weightedDrop : dropList)
            {
                float percentChance = weightedDrop.getChance() * 100;
                int count = weightedDrop.getDrop().getCount();
    
                String type = weightedDrop.getType();
                String translationKey = "exnihiloomnia.drops.type." + type;
                String translate = I18n.format(translationKey);
                
                if (translate.equals(translationKey))
                    translate = type;
                
                String drop = String.format("%.1f%% x%s Type: %s", percentChance, count, translate);
                
                toolTip.add(drop);
            }
        });
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/integration/jei/sieve/SieveRecipeCategory.kt
package com.jozufozu.exnihiloomnia.integration.jei.sieve

import com.jozufozu.exnihiloomnia.ExNihilo
import mezz.jei.api.IGuiHelper
import mezz.jei.api.gui.IDrawable
import mezz.jei.api.gui.IRecipeLayout
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.recipe.IRecipeCategory
import net.minecraft.client.Minecraft
import net.minecraft.client.resources.I18n
import net.minecraft.util.ResourceLocation

class SieveRecipeCategory(guiHelper: IGuiHelper) : IRecipeCategory<SieveRecipeWrapper> {
    private val background: IDrawable = guiHelper.createDrawable(texture, 0, 0, 166, 146)
    private var highlight: IDrawable = guiHelper.createDrawable(texture, 166, 0, 18, 18)

    override fun getUid(): String {
        return UID
    }

    override fun getTitle(): String {
        return I18n.format("jei.exnihiloomnia.sieve")
    }

    override fun getModName(): String {
        return ExNihilo.NAME
    }

    override fun getBackground(): IDrawable {
        return background
    }

    override fun drawExtras(minecraft: Minecraft) {

    }

    override fun setRecipe(recipeLayout: IRecipeLayout, sieveRecipeWrapper: SieveRecipeWrapper, ingredients: IIngredients) {
        val itemStacks = recipeLayout.itemStacks

        itemStacks.init(0, true, 74, 9)

        val outputSlots = 1

        for (slotNumber in 0..44) {
            val slotX = 2 + slotNumber % 9 * 18
            val slotY = 52 + slotNumber / 9 * 18

            itemStacks.init(outputSlots + slotNumber, false, slotX, slotY)
        }

        itemStacks.set(ingredients)

        itemStacks.addTooltipCallback { _, isInput, stack, toolTip ->
            if (isInput)
                return@addTooltipCallback

            sieveRecipeWrapper.rewards[stack]?.let {
                for ((type, dropList) in it) {
                    val translationKey = "exnihiloomnia.drops.type.$type"
                    val localizedTypeName = if (I18n.hasKey(translationKey)) I18n.format(translationKey) else type

                    toolTip.add(I18n.format("jei.exnihiloomnia.info.type_list", localizedTypeName))

                    for (weightedDrop in dropList) {
                        val percentChance = weightedDrop.chance * 100
                        val count = weightedDrop.drop.count

                        toolTip.add(I18n.format("jei.exnihiloomnia.info.reward_chance", percentChance, count))
                    }
                }
            }
        }
    }

    companion object {
        const val UID = "exnihiloomnia.sieve"

        private val texture = ResourceLocation(ExNihilo.MODID, "textures/gui/jei_sieve.png")
    }
}
